File: src/main/java/com/projectswg/launcher/core/services/data/PreferencesDataService.java
/***********************************************************************************
 * Copyright (C) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * This file is part of the ProjectSWG Launcher.                                   *
 *                                                                                 *
 * This program is free software: you can redistribute it and/or modify            *
 * it under the terms of the GNU Affero General Public License as published by     *
 * the Free Software Foundation, either version 3 of the License, or               *
 * (at your option) any later version.                                             *
 *                                                                                 *
 * This program is distributed in the hope that it will be useful,                 *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.          *
 *                                                                                 *
 ***********************************************************************************/

package com.projectswg.launcher.core.services.data;

import com.projectswg.launcher.core.resources.data.LauncherData;
import com.projectswg.launcher.core.resources.data.forwarder.ForwarderData;
import com.projectswg.launcher.core.resources.data.general.GeneralData;
import com.projectswg.launcher.core.resources.data.general.LauncherTheme;
import com.projectswg.launcher.core.resources.data.login.LoginData;
import com.projectswg.launcher.core.resources.data.login.LoginServer;
import com.projectswg.launcher.core.resources.data.update.UpdateData;
import com.projectswg.launcher.core.resources.data.update.UpdateServer;
import me.joshlarson.jlcommon.concurrency.ScheduledThreadPool;
import me.joshlarson.jlcommon.control.Service;
import me.joshlarson.jlcommon.log.Log;

import java.io.File;
import java.io.IOException;
import java.util.Locale;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

public class PreferencesDataService extends Service {
	
	private final Preferences preferences;
	private final ScheduledThreadPool executor;
	
	public PreferencesDataService() {
		this.preferences = LauncherData.INSTANCE.getPreferences();
		this.executor = new ScheduledThreadPool(1, 3, "data-executor-%d");
		loadPreferences();
	}
	
	@Override
	public boolean start() {
		createDefaults();
		executor.start();
		executor.executeWithFixedDelay(5*60000, 5*60000, this::savePreferences);
		return true;
	}
	
	@Override
	public boolean stop() {
		savePreferences();
		executor.stop();
		return executor.awaitTermination(1000);
	}
	
	private void createDefaults() {
		createPSWG();
		createTeamSWG();

		if (LauncherData.INSTANCE.getGeneral().getWine() == null || LauncherData.INSTANCE.getGeneral().getWine().isEmpty())
			LauncherData.INSTANCE.getGeneral().setWine(getWinePath());
	}
	
	private void createPSWG() {
		UpdateServer pswgUpdateServer = LauncherData.INSTANCE.getUpdate().getServers().stream().filter(s -> s.getName().equals("ProjectSWG")).findFirst().orElse(null);
		if (pswgUpdateServer == null) {
			pswgUpdateServer = new UpdateServer("ProjectSWG");
			pswgUpdateServer.setAddress("login1.projectswg.com");
			pswgUpdateServer.setPort(80);
			pswgUpdateServer.setBasePath("/launcher/patch");
			pswgUpdateServer.setGameVersion("NGE");
			LauncherData.INSTANCE.getUpdate().addServer(pswgUpdateServer);
		} else if (pswgUpdateServer.getGameVersion().isEmpty()) {
			// Migrate existing persisted update servers
			pswgUpdateServer.setGameVersion("NGE");
		}
		if (LauncherData.INSTANCE.getLogin().getServers().stream().noneMatch(s -> s.getName().equals("ProjectSWG"))) {
			LoginServer defaultLive = new LoginServer("ProjectSWG");
			defaultLive.setAddress("login1.projectswg.com");
			defaultLive.setPort(44453);
			defaultLive.setUpdateServer(pswgUpdateServer);
			LauncherData.INSTANCE.getLogin().addServer(defaultLive);
		}
		
		if (LauncherData.INSTANCE.getLogin().getServers().stream().noneMatch(s -> s.getName().equals("localhost"))) {
			LoginServer defaultLocalhost = new LoginServer("localhost");
			defaultLocalhost.setAddress("localhost");
			defaultLocalhost.setPort(44463);
			defaultLocalhost.setUpdateServer(pswgUpdateServer);
			LauncherData.INSTANCE.getLogin().addServer(defaultLocalhost);
		}
	}
	
	private void createTeamSWG() {
		UpdateServer teamswgUpdateServer = LauncherData.INSTANCE.getUpdate().getServers().stream().filter(s -> s.getName().equals("TeamSWG")).findFirst().orElse(null);
		
		if (teamswgUpdateServer == null) {
			teamswgUpdateServer = new UpdateServer("TeamSWG");
			teamswgUpdateServer.setAddress("patch.teamswg.com");
			teamswgUpdateServer.setPort(80);
			teamswgUpdateServer.setBasePath("/launcher/patch");
			teamswgUpdateServer.setGameVersion("CU");
			LauncherData.INSTANCE.getUpdate().addServer(teamswgUpdateServer);
		}
		
		if (LauncherData.INSTANCE.getLogin().getServers().stream().noneMatch(s -> s.getName().equals("Constrictor"))) {
			LoginServer constrictor = new LoginServer("Constrictor");
			constrictor.setAddress("game.teamswg.com");
			constrictor.setPort(44463);
			constrictor.setUpdateServer(teamswgUpdateServer);
			LauncherData.INSTANCE.getLogin().addServer(constrictor);
		}
	}
	
	private synchronized void loadPreferences() {
		try {
			loadGeneralPreferences(LauncherData.INSTANCE.getGeneral());
			loadUpdatePreferences(LauncherData.INSTANCE.getUpdate());
			loadLoginPreferences(LauncherData.INSTANCE.getLogin());
			loadForwarderPreferences(LauncherData.INSTANCE.getForwarderData());
		} catch (BackingStoreException e) {
			Log.w(e);
		}
	}
	
	private synchronized void savePreferences() {
		try {
			saveGeneralPreferences(LauncherData.INSTANCE.getGeneral());
			saveUpdatePreferences(LauncherData.INSTANCE.getUpdate());
			saveLoginPreferences(LauncherData.INSTANCE.getLogin());
			saveForwarderPreferences(LauncherData.INSTANCE.getForwarderData());
			preferences.flush();
		} catch (BackingStoreException e) {
			Log.w(e);
		}
	}
	
	private void loadGeneralPreferences(GeneralData generalData) {
		Preferences generalPreferences = preferences.node("general");
		ifPresent(generalPreferences, "sound", Boolean::valueOf, generalData::setSound);
		ifPresent(generalPreferences, "theme", LauncherTheme::forThemeTag, generalData::setTheme);
		ifPresent(generalPreferences, "locale", Locale::forLanguageTag, generalData::setLocale);
		ifPresent(generalPreferences, "wine", generalData::setWine);
		ifPresent(generalPreferences, "admin", Boolean::valueOf, generalData::setAdmin);
	}
	
	private void saveGeneralPreferences(GeneralData generalData) {
		Preferences generalPreferences = preferences.node("general");
		generalPreferences.putBoolean("sound", generalData.isSound());
		generalPreferences.put("theme", generalData.getTheme().getTag());
		generalPreferences.put("locale", generalData.getLocale().toLanguageTag());
		generalPreferences.putBoolean("admin", generalData.isAdmin());
		String wine = generalData.getWine();
		if (wine != null)
			generalPreferences.put("wine", wine);
	}
	
	private void loadLoginPreferences(LoginData loginData) throws BackingStoreException {
		Preferences loginPreferences = preferences.node("login");
		for (String childNodeName : loginPreferences.childrenNames()) {
			Preferences loginServerPreferences = loginPreferences.node(childNodeName);
			LoginServer server = new LoginServer(childNodeName);
			ifPresent(loginServerPreferences, "address", server::setAddress);
			ifPresent(loginServerPreferences, "port", Integer::parseInt, server::setPort);
			ifPresent(loginServerPreferences, "username", server::setUsername);
			ifPresent(loginServerPreferences, "password", server::setPassword);
			ifPresent(loginServerPreferences, "updateServer", name -> server.setUpdateServer(LauncherData.INSTANCE.getUpdate().getServers().stream().filter(s -> s.getName().equals(name)).findFirst().orElse(null)));
			ifPresent(loginServerPreferences, "verifyServer", Boolean::parseBoolean, server::setVerifyServer);
			loginData.getServers().add(server);
		}
	}
	
	private void saveLoginPreferences(LoginData loginData) throws BackingStoreException {
		preferences.node("login").removeNode();
		Preferences loginPreferences = preferences.node("login");
		for (LoginServer server : loginData.getServers()) {
			Preferences loginServerPreferences = loginPreferences.node(server.getName());
			loginServerPreferences.put("address", server.getAddress());
			loginServerPreferences.putInt("port", server.getPort());
			loginServerPreferences.put("username", server.getUsername());
			loginServerPreferences.put("password", server.getPassword());
			loginServerPreferences.putBoolean("verifyServer", server.isVerifyServer());
			UpdateServer updateServer = server.getUpdateServer();
			if (updateServer != null)
				loginServerPreferences.put("updateServer", updateServer.getName());
		}
	}
	
	private void loadUpdatePreferences(UpdateData updateData) throws BackingStoreException {
		Preferences updatePreferences = preferences.node("update");
		for (String childNodeName : updatePreferences.childrenNames()) {
			Preferences updateServerPreferences = updatePreferences.node(childNodeName);
			UpdateServer server = new UpdateServer(childNodeName);
			ifPresent(updateServerPreferences, "address", server::setAddress);
			ifPresent(updateServerPreferences, "port", Integer::parseInt, server::setPort);
			ifPresent(updateServerPreferences, "basePath", server::setBasePath);
			ifPresent(updateServerPreferences, "localPath", server::setLocalPath);
			ifPresent(updateServerPreferences, "gameVersion", server::setGameVersion);
			updateData.getServers().add(server);
		}
	}
	
	private void saveUpdatePreferences(UpdateData updateData) throws BackingStoreException {
		preferences.node("update").removeNode();
		Preferences updatePreferences = preferences.node("update");
		for (UpdateServer server : updateData.getServers()) {
			Preferences updateServerPreferences = updatePreferences.node(server.getName());
			updateServerPreferences.put("address", server.getAddress());
			updateServerPreferences.putInt("port", server.getPort());
			updateServerPreferences.put("basePath", server.getBasePath());
			updateServerPreferences.put("localPath", server.getLocalPath());
			updateServerPreferences.put("gameVersion", server.getGameVersion());
		}
	}
	
	private void loadForwarderPreferences(ForwarderData forwarderData) {
		Preferences forwarderPreferences = preferences.node("forwarder");
		int version = forwarderPreferences.getInt("version", 0);
		if (version < 1) {
			forwarderData.setSendInterval(ForwarderData.DEFAULT_SEND_INTERVAL);
			forwarderData.setSendMax(ForwarderData.DEFAULT_SEND_MAX);
		} else {
			ifPresent(forwarderPreferences, "sendInterval", Integer::valueOf, forwarderData::setSendInterval);
			ifPresent(forwarderPreferences, "sendMax", Integer::valueOf, forwarderData::setSendMax);
		}
	}
	
	private void saveForwarderPreferences(ForwarderData forwarderData) {
		Preferences forwarderPreferences = preferences.node("forwarder");
		forwarderPreferences.putInt("sendInterval", forwarderData.getSendInterval());
		forwarderPreferences.putInt("sendMax", forwarderData.getSendMax());
		forwarderPreferences.putInt("version", 1);
	}
	
	private static <T> void ifPresent(Preferences p, String key, Function<String, T> transform, Consumer<T> setter) {
		String val = p.get(key, null);
		if (val != null)
			setter.accept(transform.apply(val));
	}
	
	private static void ifPresent(Preferences p, String key, Consumer<String> setter) {
		String val = p.get(key, null);
		if (val != null)
			setter.accept(val);
	}
	
	private static String getWinePath() {
		String pathStr = System.getenv("PATH");
		if (pathStr == null)
			return null;
		
		for (String path : pathStr.split(File.pathSeparator)) {
			Log.t("Testing wine binary at %s", path);
			File test = new File(path, "wine");
			if (test.isFile()) {
				try {
					test = test.getCanonicalFile();
					Log.d("Found wine installation. Location: %s", test);
					return test.getAbsolutePath();
				} catch (IOException e) {
					Log.w("Failed to get canonical file location of possible wine location: %s", test);
				}
			}
		}
		return null;
	}
	
}

File: src/main/java/com/projectswg/launcher/services/data/PreferencesDataService.kt
/***********************************************************************************
 * Copyright (C) 2018 /// Project SWG /// www.projectswg.com                       *
 * *
 * This file is part of the ProjectSWG Launcher.                                   *
 * *
 * This program is free software: you can redistribute it and/or modify            *
 * it under the terms of the GNU Affero General Public License as published by     *
 * the Free Software Foundation, either version 3 of the License, or               *
 * (at your option) any later version.                                             *
 * *
 * This program is distributed in the hope that it will be useful,                 *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 * *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with this program.  If not, see <https:></https:>//www.gnu.org/licenses/>.          *
 * *
 */
package com.projectswg.launcher.services.data

import com.projectswg.launcher.resources.data.LauncherData
import com.projectswg.launcher.resources.data.login.AuthenticationData
import com.projectswg.launcher.resources.data.login.LoginServer
import com.projectswg.launcher.resources.data.update.UpdateServer
import me.joshlarson.jlcommon.concurrency.ScheduledThreadPool
import me.joshlarson.jlcommon.control.Service
import me.joshlarson.jlcommon.log.Log
import me.joshlarson.json.JSONInputStream
import me.joshlarson.json.JSONOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.net.URL
import java.util.*

class PreferencesDataService : Service() {
	
	private val executor: ScheduledThreadPool = ScheduledThreadPool(1, 3, "data-executor-%d")
	
	init {
		loadPreferences()
	}
	
	override fun start(): Boolean {
		createDefaults()
		executor.start()
		executor.executeWithFixedDelay((5 * 60000).toLong(), (5 * 60000).toLong()) { savePreferences() }
		return true
	}
	
	override fun stop(): Boolean {
		savePreferences()
		executor.stop()
		return executor.awaitTermination(1000)
	}
	
	private fun createDefaults() {
		if (LauncherData.INSTANCE.general.wine == null || LauncherData.INSTANCE.general.wine!!.isEmpty())
			LauncherData.INSTANCE.general.wine = winePath
	}
	
	@Synchronized
	private fun loadPreferences() {
		val applicationDataDirectory = LauncherData.getApplicationDataDirectory()
		if (!applicationDataDirectory.exists())
			applicationDataDirectory.mkdirs()
		else if (!applicationDataDirectory.isDirectory) {
			applicationDataDirectory.deleteRecursively()
			applicationDataDirectory.mkdirs()
		}
		
		lateinit var lastSelectedServerName: String
		lateinit var localServerConfiguration: Map<String, Any?>
		
		try {
			JSONInputStream(FileInputStream(File(LauncherData.getApplicationDataDirectory(), "settings.json"))).use {
				val settings = it.readObject()
				loadSettingsGeneral(settings.getMap("general"))
				loadSettingsUpdate(settings.getMap("update"))
				loadSettingsLogin(settings.getMap("login"))
				loadSettingsForwarder(settings.getMap("forwarder"))
				
				lastSelectedServerName = settings.getMap("login").getString("last_selected_server", "")
				localServerConfiguration = settings.getMap("login").getMap("local_server")
			}
		} catch (fnf: FileNotFoundException) {
			loadSettingsGeneral(mapOf())
			loadSettingsLogin(mapOf())
			loadSettingsForwarder(mapOf())
			lastSelectedServerName = ""
			localServerConfiguration = mapOf()
		}
		
		loadServers()
		
		// Add the local server last so that it's always last in the menu
		loadLocalServer(localServerConfiguration)
		
		val lastSelectedServer = LauncherData.INSTANCE.login.getServerByName(lastSelectedServerName)
		if (lastSelectedServer != null)
			LauncherData.INSTANCE.login.activeServer = lastSelectedServer
	}
	
	private fun loadSettingsGeneral(settings: Map<String, Any?>) {
		LauncherData.INSTANCE.general.locale = Locale.forLanguageTag(settings.getString("locale", Locale.US.toLanguageTag()))
		LauncherData.INSTANCE.general.wine = settings.getString("wine", "")
		LauncherData.INSTANCE.general.isAdmin = settings.getBool("admin", false)
	}
	
	private fun loadSettingsUpdate(settings: Map<String, Any?>) {
		LauncherData.INSTANCE.update.localPath = settings.getString("local_path", "")
	}
	
	private fun loadSettingsLogin(settings: Map<String, Any?>) {
		for (e in settings.getMap("authentications")) {
			val data = AuthenticationData(e.key)
			data.username = castMap(e.value).getString("username", "")
			data.password = castMap(e.value).getString("password", "")
			LauncherData.INSTANCE.login.authenticationData.add(data)
		}
	}
	
	private fun loadLocalServer(localServerSettings: Map<String, Any?>) {
		val localAuthenticationSource = AuthenticationData(LOCAL_AUTHENTICATION_NAME)
		if (LauncherData.INSTANCE.login.getAuthenticationData(LOCAL_AUTHENTICATION_NAME) == null)
			LauncherData.INSTANCE.login.authenticationData.add(localAuthenticationSource)
		
		val defaultLocalServer = LoginServer("local")
		defaultLocalServer.connectionUri = "ws://127.0.0.1:44463/game"
		defaultLocalServer.updateServer = LauncherData.INSTANCE.update.serversProperty.getOrNull(0)
		defaultLocalServer.authentication = localAuthenticationSource
		
		val localServer = loadLoginServer(localServerSettings) ?: defaultLocalServer
		LauncherData.INSTANCE.login.localServer = localServer
		LauncherData.INSTANCE.login.servers.add(LauncherData.INSTANCE.login.localServer)
	}
	
	private fun loadSettingsForwarder(settings: Map<String, Any?>) {
		LauncherData.INSTANCE.forwarder.sendInterval = settings.getInt("send_interval", 1000)
		LauncherData.INSTANCE.forwarder.sendMax = settings.getInt("send_max", 400)
	}
	
	private fun loadServers() {
		try {
			loadRemoteServers()
			JSONInputStream(FileInputStream(File(LauncherData.getApplicationDataDirectory(), "servers.json"))).use {
				val serverConfigurations = it.readObject()
				loadUpdateServers(serverConfigurations.getListOfMap("update_servers"))
				loadLoginServers(serverConfigurations.getListOfMap("login_servers"))
			}
		} catch (fnf: FileNotFoundException) {
			loadSettingsGeneral(mapOf())
			loadSettingsLogin(mapOf())
			loadSettingsForwarder(mapOf())
		}
	}
	
	private fun loadRemoteServers() {
		try {
			Log.d("Retrieving server configuration from %s", REMOTE_SERVER_CONFIGURATION_URL)
			JSONInputStream(URL(REMOTE_SERVER_CONFIGURATION_URL).openStream()).use { input ->
				JSONOutputStream(FileOutputStream(File(LauncherData.getApplicationDataDirectory(), "servers.json"))).use { output ->
					output.writeObject(input.readObject())
				}
			}
		} catch (e: IOException) {
			Log.e("Failed to retrieve updated server configuration")
		}
	}
	
	private fun loadUpdateServers(serverConfigurations: List<Map<String, Any?>>) {
		for (updateServerConfiguration in serverConfigurations) {
			val updateServer = UpdateServer(updateServerConfiguration["name"] as? String ?: continue)
			updateServer.url = updateServerConfiguration["url"] as? String ?: continue
			updateServer.gameVersion = updateServerConfiguration["game_version"] as? String ?: continue
			updateServer.friendlyName = updateServerConfiguration["friendly_name"] as? String ?: continue
			LauncherData.INSTANCE.update.servers.add(updateServer)
		}
	}
	
	private fun loadLoginServers(serverConfigurations: List<Map<String, Any?>>) {
		val unusedAuthenticationServers = HashSet(LauncherData.INSTANCE.login.authenticationData)
		for (loginServerConfiguration in serverConfigurations) {
			val loginServer = loadLoginServer(loginServerConfiguration) ?: continue
			
			unusedAuthenticationServers.remove(loginServer.authentication)
			
			LauncherData.INSTANCE.login.servers.add(loginServer)
		}
		
		// Don't want to perpetually store credentials
		for (unusedAuthentication in unusedAuthenticationServers) {
			if (unusedAuthentication.name == LOCAL_AUTHENTICATION_NAME)
				continue // This will be used later
			LauncherData.INSTANCE.login.authenticationData.remove(unusedAuthentication)
		}
	}
	
	private fun loadLoginServer(loginServerConfiguration: Map<String, Any?>): LoginServer? {
		val loginServer = LoginServer(loginServerConfiguration["name"] as? String ?: return null)
		loginServer.connectionUri = loginServerConfiguration["url"] as? String ?: return null
		loginServer.updateServer = LauncherData.INSTANCE.update.getServer(loginServerConfiguration["update_server"] as? String ?: return null) ?: return null
		val authenticationSource = loginServerConfiguration["authentication_source"] as? String ?: return null
		var authentication = LauncherData.INSTANCE.login.getAuthenticationData(authenticationSource)
		if (authentication == null) {
			authentication = AuthenticationData(loginServerConfiguration["authentication_source"] as? String ?: return null)
			LauncherData.INSTANCE.login.authenticationData.add(authentication)
		}
		loginServer.authentication = authentication
		return loginServer
	}
	
	private fun castMap(obj: Any?): Map<String, Any?> {
		@Suppress("UNCHECKED_CAST")
		return obj as? Map<String, Any?> ?: mapOf()
	}
	
	private fun castListOfMap(obj: Any?): List<Map<String, Any?>> {
		@Suppress("UNCHECKED_CAST")
		return obj as? List<Map<String, Any?>> ?: listOf()
	}
	
	@Synchronized
	private fun savePreferences() {
		JSONOutputStream(FileOutputStream(File(LauncherData.getApplicationDataDirectory(), "settings.json"))).use {
			it.writeObject(jsonSettings())
		}
	}
	
	private fun jsonSettings(): Map<String, Any?> {
		return mapOf(
			"general" to jsonSettingsGeneral(),
			"update" to jsonSettingsUpdate(),
			"login" to jsonSettingsLogin(),
			"forwarder" to jsonSettingsForwarder()
		)
	}
	
	private fun jsonSettingsGeneral(): Map<String, Any?> {
		return mapOf(
			"locale" to LauncherData.INSTANCE.general.locale.toLanguageTag(),
			"wine" to LauncherData.INSTANCE.general.wine,
			"admin" to LauncherData.INSTANCE.general.isAdmin
		)
	}
	
	private fun jsonSettingsUpdate(): Map<String, Any?> {
		return mapOf(
			"local_path" to LauncherData.INSTANCE.update.localPath
		)
	}
	
	private fun jsonSettingsLogin(): Map<String, Any?> {
		val localServer = LauncherData.INSTANCE.login.localServer
		
		return mapOf(
			"last_selected_server" to LauncherData.INSTANCE.login.lastSelectedServer,
			"authentications" to jsonSettingsLoginAuthentication(),
			"local_server" to mapOf(
				"name" to localServer.name,
				"url" to localServer.connectionUri,
				"update_server" to (localServer.updateServer?.name ?: ""),
				"authentication_source" to localServer.authentication.name
			)
		)
	}
	
	private fun jsonSettingsLoginAuthentication(): Map<String, Any?> {
		val authentications = HashMap<String, Any?>()
		
		for (auth in LauncherData.INSTANCE.login.authenticationData) {
			authentications[auth.name] = mapOf(
				"username" to auth.username,
				"password" to auth.password
			)
		}
		
		return authentications
	}
	
	private fun jsonSettingsForwarder(): Map<String, Any?> {
		return mapOf(
			"send_interval" to LauncherData.INSTANCE.forwarder.sendInterval,
			"send_max" to LauncherData.INSTANCE.forwarder.sendMax
		)
	}
	
	private fun Map<String, Any?>?.getMap(key: String): Map<String, Any?> {
		return castMap(this?.getOrDefault(key, mapOf<String, Any?>()))
	}
	
	private fun Map<String, Any?>?.getListOfMap(key: String): List<Map<String, Any?>> {
		return castListOfMap(this?.getOrDefault(key, listOf<Map<String, Any?>>()))
	}
	
	private fun Map<String, Any?>?.getString(key: String, defaultValue: String): String {
		return (this?.getOrDefault(key, defaultValue) as? String?) ?: defaultValue
	}
	
	private fun Map<String, Any?>?.getInt(key: String, defaultValue: Int): Int {
		return ((this?.getOrDefault(key, defaultValue) as? Number?) ?: defaultValue).toInt()
	}
	
	private fun Map<String, Any?>?.getBool(key: String, defaultValue: Boolean): Boolean {
		return (this?.getOrDefault(key, defaultValue) as? Boolean?) ?: defaultValue
	}
	
	companion object {
		
		private const val LOCAL_AUTHENTICATION_NAME = "local"
		private const val REMOTE_SERVER_CONFIGURATION_URL = "https://projectswg.com/launcher/servers.json"
		
		private val winePath: String?
			get() {
				val pathStr = System.getenv("PATH") ?: return null
				for (path in pathStr.split(File.pathSeparator.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
					Log.t("Testing wine binary at %s", path)
					var test = File(path, "wine")
					if (test.isFile) {
						try {
							test = test.canonicalFile
							Log.d("Found wine installation. Location: %s", test)
							return test.absolutePath
						} catch (e: IOException) {
							Log.w("Failed to get canonical file location of possible wine location: %s", test)
						}
					}
				}
				return null
			}
	}
}